File: CDA/src/main/python/programmingtheiot/cda/connection/CoapClientConnector.py
#####
# 
# This class is part of the Programming the Internet of Things project.
# 
# It is provided as a simple shell to guide the student and assist with
# implementation for the Programming the Internet of Things exercises,
# and designed to be modified by the student as needed.
#

import logging
import socket

from coapthon import defines
from coapthon.client.coap import CoAP
from coapthon.client.helperclient import HelperClient
from coapthon.messages.message import Message
from coapthon.messages.request import Request
from coapthon.utils import parse_uri
from coapthon.utils import generate_random_token

import src.main.python.programmingtheiot.common.ConfigConst as ConfigConst

from src.main.python.programmingtheiot.common.ConfigUtil import ConfigUtil
from src.main.python.programmingtheiot.common.ResourceNameEnum import ResourceNameEnum

from src.main.python.programmingtheiot.common.IDataMessageListener import IDataMessageListener
from src.main.python.programmingtheiot.cda.connection.IRequestResponseClient import IRequestResponseClient

class CoapClientConnector(IRequestResponseClient):
	"""
	Shell representation of class for student implementation.
	
	"""
	
	def __init__(self):
		self.config = ConfigUtil()
		self.dataMsgListener = None
		self.coapClient = None

		self.host = self.config.getProperty(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.HOST_KEY,
											ConfigConst.DEFAULT_HOST)
		self.port = self.config.getInteger(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.PORT_KEY,
										   ConfigConst.DEFAULT_COAP_PORT)

		logging.info('\tCoAP Server Host: ' + self.host)
		logging.info('\tCoAP Server Port: ' + str(self.port))

		self.url = "coap://" + self.host + ":" + str(self.port) + "/"

		try:
			logging.info("Parsing URL: " + self.url)

			self.host, self.port, self.path = parse_uri(self.url)
			tmpHost = socket.gethostbyname(self.host)

			if tmpHost:
				self.host = tmpHost
				self._initClient()
			else:
				logging.error("Can't resolve host: " + self.host)

		except socket.gaierror:
			logging.info("Failed to resolve host: " + self.host)
	
	def sendDiscoveryRequest(self, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		logging.info('Discovering remote resources...')

		# NOTE: we can use the API to send a discovery 'GET', or - per [RFC7252](https://tools.ietf.org/html/rfc7252#section-7.2)
		# and its reference of [RFC6690](https://tools.ietf.org/html/rfc6690#section-1.2.1), we can just send the following:
		self.coapClient.get(path='/.well-known/core', callback=self._onDiscoveryResponse, timeout=timeout)
		return True

	def sendDeleteRequest(self, resource: ResourceNameEnum, enableCON = False, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		if resource:
			logging.debug("Issuing DELETE with path: " + resource.value)
			request = self.coapClient.mk_request(defines.Codes.DELETE, path=resource.value)
			request.token = generate_random_token(2)

			if not enableCON:
				request.type = defines.Types["NON"]

			self.coapClient.send_request(request=request, callback=self._onDeleteResponse, timeout=timeout)
		else:
			logging.warning("Can't test DELETE - no path or path list provided.")

	def _onDeleteResponse(self, response):
		logging.info('DELETE response received.')

		if response:
			logging.info('Token: ' + str(response.token))
			logging.info(str(response.location_path))
			logging.info(str(response.payload))

	def sendGetRequest(self, resource: ResourceNameEnum, enableCON=False,timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		logging.info("sendGetRequest is Called");
		if resource:
			logging.debug("Issuing GET with path: " + resource.value)
			request = self.coapClient.mk_request(defines.Codes.GET, path=resource.value)
			request.token = generate_random_token(2)

			if not enableCON:
				request.type = defines.Types["NON"]

			self.coapClient.send_request(request=request, callback=self._onGetResponse, timeout=timeout)
		else:
			logging.warning("Can't test GET - no path or path list provided.")
		return False

	def _onGetResponse(self, response):
		logging.info('GET response received.')

		if response:
			logging.info('Token: ' + str(response.token))
			logging.info(str(response.location_path))
			logging.info(str(response.payload))

			#
			# NOTE: This next section is optional if you want to send a callback to the self.dataMsgListener instance
			#

			# TODO: get the URI and convert to ResourceNameEnum
			resource = None

			if self.dataMsgListener:
				self.dataMsgListener.handleIncomingMessage(resource, str(response.payload))

	def sendPostRequest(self, resource: ResourceNameEnum, payload = None, enableCON = False, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		if resource:
			"""logging.debug("Issuing POST with path: " + resource.value)"""
			request = self.coapClient.mk_request(defines.Codes.POST, path=resource.value)
			request.token = generate_random_token(2)
			request.payload = payload

			if not enableCON:
				request.type = defines.Types["NON"]

			self.coapClient.send_request(request=request, callback=self._onPostResponse, timeout=timeout)
		else:
			"""logging.warning("Can't test POST - no path or path list provided.")"""

	def _onPostResponse(self, response):
		"""logging.info('POST response received.')"""

		if response:
			logging.info('Token: ' + str(response.token))
			logging.info(str(response.location_path))
			logging.info(str(response.payload))

	def sendPutRequest(self, resource: ResourceNameEnum, payload = None, enableCON = False, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		if resource:
			"""logging.debug("Issuing PUT with path: " + resource.value)"""
			request = self.coapClient.mk_request(defines.Codes.PUT, path=resource.value)
			request.token = generate_random_token(2)
			request.payload = payload

			if not enableCON:
				request.type = defines.Types["NON"]

			self.coapClient.send_request(request=request, callback=self._onPutResponse, timeout=timeout)
		else:
			"""logging.warning("Can't test PUT - no path or path list provided.")"""

	def _onPutResponse(self, response):
		"""logging.info('PUT response received.')"""

		if response:
			logging.info('Token: ' + str(response.token))
			logging.info(str(response.location_path))
			logging.info(str(response.payload))

	def setDataMessageListener(self, listener: IDataMessageListener) -> bool:
		self.dataMsgListener = listener
		return True

	def startObserver(self, resource: ResourceNameEnum, ttl: int = IRequestResponseClient.DEFAULT_TTL) -> bool:
		pass

	def stopObserver(self, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
		pass
	
	def _initClient(self):
		if not self.coapClient:
			self.coapClient = HelperClient(server=(self.host, self.port))

	def _onDiscoveryResponse(self, response):
		if response:
			logging.info(response.pretty_print())

			# get the payload and convert to a list of paths
			if response.payload is None:
				logging.info('  Path entry ['']:' + 'None')
				return
			self.pathList = response.payload.split(',')
			index = 0

			# the following is optional, but provides an easy way to track all the returned resource names
			for path in self.pathList:
				for char in '<\>':
					path = path.replace(char, '')

				self.pathList[index] = path

				logging.info('  Path entry [' + str(index) + ']:' + self.pathList[index])
				index += 1
		else:
			logging.info("No response received.")

File: GDA/src/main/java/programmingtheiot/gda/connection/CoapClientConnector.java
/**
 * This class is part of the Programming the Internet of Things project.
 * 
 * It is provided as a simple shell to guide the student and assist with
 * implementation for the Programming the Internet of Things exercises,
 * and designed to be modified by the student as needed.
 */ 

package programmingtheiot.gda.connection;

import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.eclipse.californium.core.CoapClient;
import org.eclipse.californium.core.CoapHandler;
import org.eclipse.californium.core.CoapResponse;
import org.eclipse.californium.core.WebLink;
import org.eclipse.californium.core.coap.MediaTypeRegistry;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.gda.connection.handlers.GenericCoapResponseHandler;

/**
 * Shell representation of class for student implementation.
 *
 */
public class CoapClientConnector implements IRequestResponseClient
{
	// static
	
	private static final Logger _Logger =
		Logger.getLogger(CoapClientConnector.class.getName());
	
	// params
	private String     protocol;
	private String     host;
	private int        port;
	private String     serverAddr;
	private CoapClient clientConn;
	private IDataMessageListener dataMsgListener;
	
	// constructors
	
	/**
	 * Default.
	 * 
	 * All config data will be loaded from the config file.
	 */
	public CoapClientConnector() {
		ConfigUtil config = ConfigUtil.getInstance();
		this.host = config.getProperty(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.HOST_KEY, ConfigConst.DEFAULT_HOST);

		if (config.getBoolean(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.ENABLE_CRYPT_KEY)) {
			this.protocol = ConfigConst.DEFAULT_COAP_SECURE_PROTOCOL;
			this.port     = config.getInteger(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.SECURE_PORT_KEY, ConfigConst.DEFAULT_COAP_SECURE_PORT);
		} else {
			this.protocol = ConfigConst.DEFAULT_COAP_PROTOCOL;
			this.port     = config.getInteger(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.PORT_KEY, ConfigConst.DEFAULT_COAP_PORT);
		}
		// NOTE: URL does not have a protocol handler for "coap",
// so we need to construct the URL manually
		// NOTE: URL does not have a protocol handler for "coap",
// so we need to construct the URL manually
		this.serverAddr = this.protocol + "://" + this.host + ":" + this.port;

		initClient();

		_Logger.info("Using URL for server conn: " + this.serverAddr);
	}

	public CoapClientConnector(String host, boolean isSecure, boolean enableConfirmedMsgs)
	{

	}


	// public methods
	
	@Override
	public boolean sendDiscoveryRequest(int timeout)
	{
		_Logger.info("sendDiscoveryRequest is called.");
		try {
			//clientConn = new CoapClient("coap://localhost:5683");
			Set<WebLink> wlSet = this.clientConn.discover();

			if (wlSet != null) {
				for (WebLink wl : wlSet) {
					_Logger.info(" --> URI: " + wl.getURI() + ". Attributes: " + wl.getAttributes());
				}
			}
			return true;
		}catch(Exception ex) {
			_Logger.info(("There is an error!"));
			return false;
		}





	}
	@Override
	public boolean sendDeleteRequest(ResourceNameEnum resource,int timeout)
	{
		_Logger.info("sendDeleteRequest is called.");

		return false;

	}

	@Override
	public boolean sendGetRequest(ResourceNameEnum resource, int timeout)
	{
		_Logger.info("sendGetRequest is called.");
		return false;
	}




	@Override
	public boolean sendPostRequest(ResourceNameEnum resource, String payload, int timeout)
	{
		//_Logger.info("sendPostRequest is called.");

		return false;
	}

	@Override
	public boolean sendPutRequest(ResourceNameEnum resource, String payload, int timeout)
	{
		//_Logger.info("sendPutRequest is called.");

		return false;
	}


	@Override
	public boolean setDataMessageListener(IDataMessageListener listener)
	{
		this.dataMsgListener = listener;
		return true;
	}

	@Override
	public boolean startObserver(ResourceNameEnum resource, int ttl)
	{
		return false;
	}

	@Override
	public boolean stopObserver(int timeout)
	{
		return false;
	}

	public boolean sendDeleteRequest(ResourceNameEnum resource, boolean enableCON, int timeout) {
		CoapResponse response = null;

		if (enableCON) {
			this.clientConn.useCONs();
		} else {
			this.clientConn.useNONs();
		}

		this.clientConn.setURI(this.serverAddr + "/" + resource.getResourceName());
		response = this.clientConn.delete();

		if (response != null) {
			// TODO: implement your logic here

			_Logger.info("Handling DELETE. Response: " + response.isSuccess() + " - " + response.getOptions() + " - " +
					response.getCode() + " - " + response.getResponseText());

			if (this.dataMsgListener != null) {
				// TODO: implement this
			}

			return true;
		} else {
			_Logger.warning("Handling DELETE. No response received.");
		}

		return false;
	}
	public boolean sendPostRequest(ResourceNameEnum resource, boolean enableCON, String payload, int timeout) {
		CoapResponse response = null;

		if (enableCON) {
			this.clientConn.useCONs();
		} else {
			this.clientConn.useNONs();
		}

		this.clientConn.setURI(this.serverAddr + "/" + resource.getResourceName());

// TODO: determine which MediaTypeRegistry const should be used for this call
		response = this.clientConn.post(payload, MediaTypeRegistry.TEXT_PLAIN);

		if (response != null) {
			// TODO: implement your logic here

			_Logger.info("Handling POST. Response: " + response.isSuccess() + " - " + response.getOptions() + " - " +
					response.getCode() + " - " + response.getResponseText());

			if (this.dataMsgListener != null) {
				// TODO: implement this
			}

			return true;
		} else {
			_Logger.warning("Handling POST. No response received.");
		}

		return false;
	}
	public boolean sendPutRequest(ResourceNameEnum resource, boolean enableCON, String payload, int timeout) {
		CoapResponse response = null;

		if (enableCON) {
			this.clientConn.useCONs();
		} else {
			this.clientConn.useNONs();
		}

		this.clientConn.setURI(this.serverAddr + "/" + resource.getResourceName());

// TODO: determine which MediaTypeRegistry const should be used for this call
		response = this.clientConn.put(payload, MediaTypeRegistry.TEXT_PLAIN);

		if (response != null) {
			// TODO: implement your logic here

			_Logger.info("Handling PUT. Response: " + response.isSuccess() + " - " + response.getOptions() + " - " +
					response.getCode() + " - " + response.getResponseText());

			if (this.dataMsgListener != null) {
				// TODO: implement this
			}

			return true;
		} else {
			_Logger.warning("Handling PUT. No response received.");
		}

		return false;
	}

	public boolean sendGetRequest(ResourceNameEnum resource, boolean enableCON, int timeout) {
		CoapResponse response = null;

		if (enableCON) {
			this.clientConn.useCONs();
		} else {
			this.clientConn.useNONs();
		}

		this.clientConn.setURI(this.serverAddr + "/" + resource.getResourceName());
		response = this.clientConn.get();

		if (response != null) {
			// TODO: implement your logic here

			_Logger.info("Handling GET. Response: " + response.isSuccess() + " - " + response.getOptions() + " - " +
					response.getCode() + " - " + response.getResponseText());

			if (this.dataMsgListener != null) {
				// TODO: implement this
			}

			return true;
		} else {
			_Logger.warning("Handling GET. No response received.");
		}

		return false;
	}
	
	// private methods
	private void initClient() {
		try {
			this.clientConn = new CoapClient(this.serverAddr);

			_Logger.info("Created client connection to server / resource: " + this.serverAddr);
		} catch (Exception e) {
			_Logger.log(Level.SEVERE, "Failed to connect to broker: " + (this.clientConn != null ? this.clientConn.getURI() : this.serverAddr), e);
		}
	}
}
